File: 0093-restore-ip-addresses/0093-restore-ip-addresses.java
class Solution {
    private StringBuilder path;
    private List<String> ans = new ArrayList<>();
    
    public List<String> restoreIpAddresses(String s) {
        path = new StringBuilder(s);
        backtracking(path, 0, 0);
        return ans;
    }
    
    private void backtracking(StringBuilder path, int points, int startIdx) {
        if (points == 3) {
            if (isValid(path, startIdx, path.length() - 1)) {
                ans.add(path.toString());
            }
            return;
        }
        
        for (int i = startIdx; i < path.length(); i++) {
            if (!isValid(path, startIdx, i)) {
                break;
            }
            path.insert(i + 1, ".");
            backtracking(path, points + 1, i + 2);
            path.deleteCharAt(i + 1);
        }
    }
    
    private boolean isValid(StringBuilder path, int i, int j) {
        if (i >= path.length()) {
            return false;
        }
        
        if (path.charAt(i) == '0' && i < j) {
            return false;
        }
        
        if (j - i + 1 > 3) {
            return false;
        }
        
        if (Integer.parseInt(path.substring(i, j + 1)) > 255) {
            return false;
        }
        
        return true;
    }
}

File: 0093-restore-ip-addresses/0093-restore-ip-addresses.py
class Solution:
    def restoreIpAddresses(self, s: str) -> List[str]:
        ans = []

        def isValid(s: str) -> bool:
            return len(s) == 1 or s[0] != "0" and int(s) <= 255
        
        def backtracking(index: int, parts: List[str]) -> None:
            if len(parts) == 4 and index == len(s):
                ans.append(".".join(parts))
                return
            
            for i in range(index, len(s)):
                substring = s[index: i + 1]
                if isValid(substring):
                    parts.append(substring)
                    backtracking(i + 1, parts)
                    parts.pop()

        backtracking(0, [])
        return ans
